File: src/js/sql_ebind.js
/* sql_ebind
 *
 * Enhanced name binding for SQL queries implemented in js
 *
 * see sql_ebind doc on GitHub
 *
 * @param string $sql The sql query with params to be bound
 * @param Array	$bind	Assoc array of params to bind
 * @param string $bind_marker what to use
 *
 * @return Object(string sql with names replaced, array of normal params)
 */
function sql_ebind(sql, bind, bind_marker) {
	'use strict';

	// handle default values
	bind = (typeof bind !== 'undefined') ? bind : [];
	bind_marker = (typeof bind_marker !== 'undefined') ? bind_marker : '?';

	// to hold $pattern matches
	var bind_matches = [];
	// to hold ordered list of replacements
	var ord_bind_list = [];
	// limit to catch endless replacement loops
	var loop_limit = 1000;

	// Bind abnormal params
	// straight string substitution, don't add anything to $ord_bind_list
	var pattern;

	// prevend endless replacement
	var repeat = 0;
	var i;
	var matches, matches_length, field;

	var val, subs;

	// loop over $bind once to gather hashes for phases
	var phase1 = [];
	var phase2 = [];
	var phase3 = [];
	for (var key in bind) {
		if (typeof bind[key] !== 'function')  // key is an object
			continue;
    val = bind[key];
    if (key.substring(0, 4) == '{{{:') { // filename
      phase1[key] = val;
    }
    else if (key.substring(0, 3) == '{{:') {  // structural param
      phase2[key] = val;
    }
      else if (key.substring(0, 2) == '{:') {  // normal param
        phase3[key] = val;
      }
    // else there's junk in $bind
    }

	// Phase 1:	inline replace from file
	// Bind included files
	/*
	subs = 0;
	do {
		if (repeat++ > loop_limit) {
			throw arguments.callee.name + ' repeat limit reached, check params for circular references';
		}
		for (var key in phase1) {
			val = phase1[key];
			if (typeof val !== 'function') {
				// here's where hand-waving occurs
				// server-side will have its way
				// client-side will have its way
				replacewith = somehow get file contents;
				sql = sql.replace(bind_matches[i], phase1[bind_matches[i]]);
				subs++;
		}
	} while (subs > 0)
	*/

	// Phase 2:	Bind abnormal params
	// iterate until no more double-curly-bracket expressions in $sql
	repeat = 0;
	do {
		if (repeat++ > loop_limit) {
			throw arguments.callee.name + ' repeat limit reached, check params for circular references';
		}
		for (key in phase2) {
			val = phase2[key];
			if ( typeof val !== 'function'  // key is not an object
				&& sql.indexOf(key) > -1   // key is in sql
			) {
				sql = sql.replace(key, bind[key]);
				subs++;
			}
		}
	} while (subs > 0);

	// Phase 3:	Bind normal params
	pattern = new RegExp("{:[A-Za-z][A-Za-z0-9_]*}");
	bind_matches = [];
	repeat = 0;
	// iterate until no more single-curly-bracket expressions in $sql
	do {
		matches_length = 0;
		if (repeat++ > loop_limit) {
			throw arguments.callee.name + ' repeat limit reached, check params for circular references';
		}
		matches = sql.match(pattern);
		if (matches) {
			matches_length = matches.length;
		}
		if (matches_length > 0) {
			for (i=0; i<matches.length; i++) {
				bind_matches[i] = (matches[i]).trim();
			}
			for (i=0; i<bind_matches.length; i++) {
				field = bind_matches[i];
				if (bind[field].isArray) {
					sql = sql.replace(
						new RegExp(bind_matches[i]),
						Array(bind[field].length)
							.fill(bind_marker)
							.join(', ')
					);
					ord_bind_list = ord_bind_list.concat(bind[field]);
				} else {
					sql = sql.replace(bind_matches[i], bind_marker);
					ord_bind_list.push(bind[field]);
				}
			}
		}
	} while (matches_length > 0);

	return {'sql': sql, 'params': ord_bind_list};
}

File: src/python/sql_ebind.py
import re
import sys
import os

def sql_ebind(sql, bind = {}, bind_marker = '?'):
    """
    sql_ebind

    see sql_ebind doc on GitHub

    Args:
        sql (string): The sql query with params to be bound
        bind (Dict): Dict of params to bind
        bind_marker (str): string to use for parameter placeholder

    Returns:
        Dict: {
            sql:    (string) ready-to-prepare SQL, 
            params: (List)   ordered list of params for sql
        }
    """

    # to hold pattern matches
    bind_matches = {}
    # to hold ordered list of replacements
    ord_bind_list = []
    # limit to catch endless replacement loops
    loop_limit = 1000
    subs = 0

    # loop over bind once to gather hashes for phases
    phase1 = {}
    phase2 = {}
    phase3 = {}
    for key in bind.iterkeys():
        val = bind[key]
        if (type(val) != 'function'):  # key is an object
            continue
        if (key.index('{{{:') == 0): # filename
            phase1[key] = val
        elif (key.index('{{:') == 0):  # structural param
            phase2[key] = val
        elif (key.index('{:') == 0):  # normal param
            phase3[key] = val
        # else there's junk in bind

    # Phase 1:  inline replace from file
    repeat = 0
    while True:
        if repeat > loop_limit:
            raise ValueError(sys._getframe().f_code.co_name + ' repeat limit reached, check params for circular references')
        subs = 0
        for key in phase1.iterkeys():
            val = phase1[key]
            if (sql.index(key) > -1):
                filename = val
                contents = ''
                if (os.path.isfile(filename)):
                    with open(filename) as f:
                        contents = f.read()
                else:
                    raise ValueError('Unable to locate ' + os.getcwd() + '/' + filename)
                sql = sql.replace(key, contents)
                subs += 1

    # Phase 2:  Bind abnormal params, such as structural objects
    repeat = 0
    while True:
        if repeat > loop_limit:
            raise ValueError(sys._getframe().f_code.co_name + ' repeat limit reached, check params for circular references')
        subs = 0
        for key in phase2.iterkeys():
            val = phase2[key]
            if (sql.index(key) > -1):
                sql = sql.replace(key, val)
                subs += 1

    # Phase 3:  Bind normal params
    pattern = re.compile("{:[A-Za-z][A-Za-z0-9_]*}")
    bind_matches = []
    repeat = 0
    # iterate until no more single-curly-bracket expressions in sql
    while True:
        matches_length = 0
        repeat += 1
        if repeat > loop_limit:
            raise ValueError(sys._getframe().f_code.co_name + ' repeat limit reached, check params for circular references')
        matches = re.findall(pattern, sql)
        matches_length = len(matches)
        for i in range(len(matches)):
            match = matches[i]
            bind_matches.append(match)
            field = bind_matches[i]
            if type(phase3[field]) == list:
                sql = re.sub(bind_matches[i], (', ').join([bind_marker] * phase3[field].length), sql)
                ord_bind_list = ord_bind_list + phase3[field]
            else:
                sql = re.sub(bind_matches[i], bind_marker, sql)
                ord_bind_list.append(phase3[field])
            subs += 1
        if (matches_length == 0):
            break

    return {'sql': sql, 'params': ord_bind_list}
